Update to Python 3.11

seanbudd opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review". 
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Closes #12064 
Closes #12551 
Closes #15577 
Closes #15167

### Summary of the issue:
Python needs to be updated to 3.11, as Python 3.7 is EOL.
Python pip dependencies need to be updated to match the python upgrade.
`typing_extensions` is no longer needed.

### Description of user facing changes
Performance and security enhancements from dependency upgrades.

### Description of development approach
- Updates python in build scripts
- Updates pip dependencies
- Set Windows 8.1 (Blue) as minimum windows version
- Update references to python and windows version in docs
- nose is replaced by the unittest module with xmlrunner to generate XML test output
- drop python optimization to level 0 from level 1. level 1 removes `asserts`.

### Testing strategy:
Alpha testing required

### Known issues with pull request:
* BrlAPI / BrlTTY will be broken until https://github.com/nvaccess/nvda-misc-deps/pull/27 is merged and updated in master
* There is an intermittent issue copying text in Chrome https://github.com/nvaccess/nvda/pull/15544#issuecomment-1754154598
* NVDA's garbageHandler is more sensitive https://github.com/nvaccess/nvda/pull/15544#issuecomment-1758998862



### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

seanbudd committed:
Update to Python 3.11

seanbudd committed:
update changes

codeofdusk commented:
This is excellent to see!

* Py2exe changed their API, so building a binary distribution won't be possible without [significant updates](https://github.com/py2exe/py2exe/blob/master/docs/migration.md).
* This branch runs very well under Python 3.12, slated for release on Monday. NVDA's upgrade is blocked by py2exe/py2exe#191, though.

codeofdusk commented on user_docs/en/changes.t2t:
```diff
@@ -28,9 +28,17 @@ Please refer to [the developer guide https://www.nvaccess.org/files/nvda/documen
-- Updated Comtypes to version 1.2.0. (#15513)
-- Updated fast_diff_match_patch to version 2.0.1. (#15514)
-- Updated SCons to version 4.5.2. (#15529)
+- NVDA is now built with Python 3.11 (#12064)
```
Style consistency:

```suggestion
- Updated Python to 3.11 (#12064)
```

seanbudd replied:
These are separate items. 
Python is part of the build process, it's not an internal project dependency. We haven't updated a version of Python internally to the project, we've updated what we use to build the project.

codeofdusk replied:
Makes sense. Both previous Python build-time updates use the "updated Python to x" language.

seanbudd commented:
We have no plan to upgrade to 3.12 in the 2024.1 release cycle

seanbudd committed:
fix py2exe

leonardder commented on nvdaHelper/readme.md:
```diff
@@ -70,7 +70,7 @@ You should still build on the command line to verify errors.
 -  Specify Debug Configuration Settings
   - Build command line: `scons source`
-  - Preprocessor definitions (/D): `WIN32;_WINDOWS;_USRDLL;NVDAHELPER_EXPORTS;UNICODE;_CRT_SECURE_NO_DEPRECATE;LOGLEVEL=15;_WIN32_WINNT=_WIN32_WINNT_WIN7;NOMINMAX`
+  - Preprocessor definitions (/D): `WIN32;_WINDOWS;_USRDLL;NVDAHELPER_EXPORTS;UNICODE;_CRT_SECURE_NO_DEPRECATE;LOGLEVEL=15;_WIN32_WINNT=_WIN32_WINNT_WINBLUE;NOMINMAX`
```
Interesting enough, the readme is updated but archBuild_sconscript seems not.

When updating archBuild_sconscript, also consider updating the linker subsystem, I guess that can become 6.03 now, though it is very hard to track down documentation on this.

seanbudd replied:
ah it got missed being git added at some point.

Can you explain what you mean by updating the linker subsystem?

leonardder replied:
Sure. search for `subsystem = "/subsystem:windows,6.02"`

seanbudd replied:
No updates seem available according to windows docs

https://learn.microsoft.com/en-us/cpp/build/reference/subsystem-specify-subsystem?view=msvc-170#remarks

seanbudd replied:
nevermind

seanbudd replied:
fixed both of these issues raised

LeonarddeR reviewed (commented):
Really love to see this flying!
Before this is merged, I'd like to suggest changing the launcher nsis script so that it has Windows 8.1 as the minimum required version as well.

codeofdusk commented on requirements.txt:
```diff
@@ -4,28 +4,21 @@ SCons==4.5.2
-wxPython==4.1.1
-git+https://github.com/DiffSK/configobj@3e2f4cc#egg=configobj
+wxPython==4.2.1
+configobj==5.0.8
```
There have been a few further commits on configobj `master` (5.1), but mostly typo fixes. The project is looking for a new maintainer.

seanbudd replied:
changed to the latest commit rather than a pinned version

seanbudd committed:
fix compile

seanbudd committed:
update linker subsystem

lukaszgo1 commented on requirements.txt:
```diff
@@ -4,28 +4,26 @@ SCons==4.5.2
-wxPython==4.1.1
-git+https://github.com/DiffSK/configobj@3e2f4cc#egg=configobj
+wxPython==4.2.1
+git+https://github.com/DiffSK/configobj@e2ba445#egg=configobj
```
While at it could you please use full hash of the commit? If incomplete hash is used pip shows the following warning:
```
  WARNING: Did not find branch or tag '3e2f4cc', assuming revision or ref.
```

seanbudd replied:
fixed

lukaszgo1 commented on requirements.txt:
```diff
@@ -35,6 +33,6 @@ flake8-tabs == 2.1.0
 nose==1.3.7
```
The version of Nose we're using is incompatible with Python 3.11, and it seems no new version is available. [Nose3](https://pypi.org/project/nose3/) is an alternative, though I'm not sure to what extend it can be used as a drop-in replacement.

seanbudd replied:
Nose3 doesn't support python 3.11. Nose2 is an alternative but I think dropping nose for native unittesting and moving to pytest if required is ideal.

seanbudd replied:
Nose2 doesn't explicitly support 3.11 either

seanbudd replied:
I've dropped nose as a dependency. It seems like we no longer need its test discovery

lukaszgo1 commented on requirements.txt:
```diff
@@ -4,28 +4,26 @@ SCons==4.5.2
-
-# Requirements for type checking.
-# typing_extensions is incorporated in py3.8+, also available via mypy
-typing_extensions==4.3.0
```
Can the fact that we no longer bundle `typing_extensions` be mentioned in the Changes for developers section? Some add-ons may import from it.

seanbudd replied:
done, also referenced nose being dropped

lukaszgo1 commented on source/setup.py:
```diff
@@ -110,80 +92,64 @@ def getRecursiveDataFiles(dest,source,excludes=()):
-			"description": "NVDA Diff-match-patch proxy",
-			"product_name": name,
-			"product_version": version,
-			"copyright": f"{NVDAcopyright}, Bill Dengler",
```
Previously nvda_dmp's binary had different copyright info than all other binaries. While this is obviously a  small think, and if it is difficult to support with the latest py2exe not  very important, just wanted to make sure this change was intentional.

lukaszgo1 replied:
It turns out I was mistaken - we used to specify different version info for most of the binaries and not only for nvda_dmp. It seems this cannot be currently done with the new freeze interface to py2exe, as it accepts a single dictionary with version data in the freeze function. Would it be useful to report this to them, and perhaps invest in creating a fix?

seanbudd replied:
I'm assuming this is an intended change that the developers don't intend to fix, but it may be worth reporting

lukaszgo1 replied:
Looking at the discussion in which the new freeze API was introduced I don't think this was done on purpose. I have created https://github.com/py2exe/py2exe/issues/193 to discuss if this can be reintroduced.

lukaszgo1 replied:
It is now once again possible to specify different version info for each of the binaries in py2exe. Unfortunately while my fix is on py2exe's master, I have missed the release by a few hours. We can either note the fact that version info has changed as a known issue of this PR, and fix it up whenever next version of py2exe is released, or install it from git for the time being. Commit 4e7b2b2c60face592e67cb1bc935172a20fa371d, contains the fix but is otherwise identical to the latest official release.

seanbudd replied:
Thanks, this should be fixed now using that commit

lukaszgo1 commented on user_docs/en/userGuide.t2t:
```diff
@@ -39,9 +39,7 @@ Major highlights include:
-- Operating Systems: all 32-bit and 64-bit editions of Windows 7, Windows 8, Windows 8.1, Windows 10, Windows 11, and all Server Operating Systems starting from Windows Server 2008 R2.
-  - For Windows 7, NVDA requires Service Pack 1 or higher.
-  - For Windows Server 2008 R2, NVDA requires Service Pack 1 or higher.
+- Operating Systems: all 32-bit and 64-bit editions of Windows 8.1, Windows 10, Windows 11, and all Server Operating Systems starting from Windows Server 2012.
```
`winVersion.isSupportedOS` should also be updated. Note that doing so might be slightly tricky, as we need to differentiate between client version of Windows (where Windows 8.1 is NT 6.3), and server (where Server 2012 is NT 6.2).

leonardder replied:
Actually, isn't this a mistake, i.e. shouldn't the minimum server os be 2012R2?

```suggestion
- Operating Systems: all 32-bit and 64-bit editions of Windows 8.1, Windows 10, Windows 11, and all Server Operating Systems starting from Windows Server 2012 R2.
```

seanbudd replied:
That's correct, it should be R2 and Win 6.3.

seanbudd commented:
>Before this is merged, I'd like to suggest changing the launcher nsis script so that it has Windows 8.1 as the minimum required version as well.

As far as I am aware nvdaLauncher.nsi doesn't check Windows versions

seanbudd committed:
fixup isSupportedOS

seanbudd committed:
drop nose as dependency

seanbudd committed:
fix server 2012 r2 ref

seanbudd committed:
use xmlrunner for xml unit test output

seanbudd committed:
fix unit test report name

seanbudd committed:
fix abstract method unit test regex changes

seanbudd committed:
fix lint

seanbudd committed:
remove unused keyword

seanbudd committed:
upgrade optimization level to 2

seanbudd committed:
fix uiaccess parsing

seanbudd commented:
Note, this is nearing being stable for testing. The only known issues left are:
- updating BRLAPI
- fixing the system tests from failing (this will require some investigation).

codeofdusk commented:
@seanbudd Performance in MSAA applications is extremely poor when running on a binary copy. This performance degredation is not experienced when running from source.

STR:
1. `scons dist` (with a certificate for UI Access).
2. run `nvda_uiaccess.exe`, install the built version (NVDA menu -> tools -> install NVDA).
3. When installation completes, press <kbd>Win</kbd>+<kbd>r</kbd>.
4. Tab around. On my system, I experience over a second of lag.

seanbudd commented:
I've also experienced some lag. Was this occurring before commit 0d6970bf785ee7dd831230c786e14291a0a91365 ?

codeofdusk commented:
0d6970b did not affect this lag.

seanbudd commented:
Thanks - we should update our MSAA interfaces in https://github.com/nvaccess/nvda-misc-deps/pull/27 which may improve this. Note the current PR is not in a buildable state.

Otherwise, this may be tricky to hunt down.

leonardder commented:
> Thanks - we should update our MSAA interfaces in [nvaccess/nvda-misc-deps#27](https://github.com/nvaccess/nvda-misc-deps/pull/27) which may improve this. Note the current PR is not in a buildable state.
> 
> Otherwise, this may be tricky to hunt down.

I think it is very unliekly that updating MSAA interfaces will fix this.
it might be an issue in py2exe that requires us to test older versions.

lukaszgo1 commented:
I am also experiencing these performance issues. The cause appears to be as follows:
From the log of a frozen NVDA (threads other than the main one omitted for brevity, as they do not contribute anything useful to the picture):
```
Python stack for thread 1140 (MainThread):
  File "nvda.pyw", line 399, in <module>
  File "core.pyc", line 813, in main
  File "wx\core.pyc", line 2262, in MainLoop
  File "wx\core.pyc", line 3427, in <lambda>
  File "core.pyc", line 114, in doStartupDialogs
  File "gui\startupDialogs.pyc", line 145, in run
  File "wx\core.pyc", line 3427, in <lambda>
  File "core.pyc", line 746, in processRequest
  File "core.pyc", line 762, in Notify
  File "queueHandler.pyc", line 97, in pumpAll
  File "queueHandler.pyc", line 64, in flushQueue
  File "eventHandler.pyc", line 84, in _queueEventCallback
  File "eventHandler.pyc", line 322, in executeEvent
  File "eventHandler.pyc", line 355, in doPreGainFocus
  File "api.pyc", line 144, in setFocusObject
  File "baseObject.pyc", line 62, in __get__
  File "baseObject.pyc", line 168, in _getPropertyViaCache
  File "NVDAObjects\__init__.pyc", line 656, in _get_container
  File "baseObject.pyc", line 62, in __get__
  File "baseObject.pyc", line 168, in _getPropertyViaCache
  File "NVDAObjects\IAccessible\__init__.pyc", line 1061, in _get_parent
  File "IAccessibleHandler\__init__.pyc", line 471, in accParent
  File "comtypes\client\__init__.pyc", line 47, in wrap_outparam
  File "comtypes\client\__init__.pyc", line 115, in GetBestInterface
  File "comtypes\client\_generate.pyc", line 140, in GetModule
  File "comtypes\client\_generate.pyc", line 207, in _create_friendly_module
  File "monkeyPatches\comtypesMonkeyPatches.pyc", line 192, in new_my_import
  File "importlib\__init__.pyc", line 70, in invalidate_caches
  File "<frozen importlib._bootstrap_external>", line 1403, in invalidate_caches
  File "<frozen zipimport>", line 333, in invalidate_caches
  File "<frozen zipimport>", line 524, in _read_directory

```

NVDA freezes in `importlib.invalidate_caches` called by one of our monkey patches to comtypes. I've confirmed that executing `importlib.invalidate_caches` from the Python console of the binary build created with Python 3.11 takes a long time, whereas in the Python 3.7 based builds it is almost instantaneous. That is probably because in Python 3.8 `zipimport` was rewritten in Python for easier maintainability (it used to be written in C). Thankfully the monkey patch to comtypes which is causing us troubles appears to be no longer needed. It was introduced in #10235 to fix #10228 as a stop-gap solution, since the version of comtypes we we're using back then had no fix for this integrated. This is now fixed in https://github.com/enthought/comtypes/pull/172, and the fix is included in a released version of comtypes. note that it would not reintroduce the same issues, since even though it also relies on `importlib.invalidate_caches`, it executes  it only when creating a new module, and not on every import of a com interface. In short I suggest to remove `replace_my_import` and associated functions from our monkey patches for comtypes. On my system with a binary build the performance is back to normal after that, however I haven't tested explicitly with a signed installed binary build. @codeofdusk Would you be able to do this?

codeofdusk commented:
@lukaszgo1 Removing `replace_my_import` restores MSAA performance on my system. Thanks for tracking it down!

seanbudd committed:
use round not int

seanbudd committed:
fixup configobj changes ref

seanbudd committed:
Merge remote-tracking branch 'origin/master' into py311-min

codeofdusk commented on user_docs/en/changes.t2t:
```diff
@@ -28,16 +28,28 @@ Please refer to [the developer guide https://www.nvaccess.org/files/nvda/documen
 These are breaking API changes.
 Please open a GitHub issue if your Add-on has an issue with updating to the new API.
 
+- NVDA is now built with Python 3.11 (#12064)
```
```suggestion
- Updated Python to 3.11. (#12064)
```

seanbudd replied:
You made this suggestion previously here https://github.com/nvaccess/nvda/pull/15544#discussion_r1339474402 - it appears to be deleted now. The reasoning against it still applies. We wouldn't say "Updated to Visual Studio 2022", we would say "NVDA is now built with VS 2022".

codeofdusk replied:
In the case of VS, this affects compiled (i.e. NVDAHelper) code, not interpreted Python run from an included DLL. Also we have used the "updated Python to x" language with all other updates (see 2019.3 and 2015.1 changes for developers).

codeofdusk commented on user_docs/en/changes.t2t:
```diff
@@ -28,16 +28,28 @@ Please refer to [the developer guide https://www.nvaccess.org/files/nvda/documen
 
+- NVDA is now built with Python 3.11 (#12064)
+- Updated pip dependencies:
+  - configobj to 5.0.8. (#15544)
```
```suggestion
  - configobj to 5.1.0dev commit e2ba445. (#15544)
```

seanbudd replied:
This has been resolved in ef044ae

codeofdusk commented on user_docs/en/changes.t2t:
```diff
@@ -28,9 +28,17 @@ Please refer to [the developer guide https://www.nvaccess.org/files/nvda/documen
-- Updated Comtypes to version 1.2.0. (#15513)
-- Updated fast_diff_match_patch to version 2.0.1. (#15514)
-- Updated SCons to version 4.5.2. (#15529)
+- NVDA is now built with Python 3.11 (#12064)
```
> it's not an internally bundled dependency

How not? Py2exe bundles a Python interpreter dll as part of the build process, and when developing NVDA runs using the system installed interpreter. Strictly speaking "updated python to x" is more accurate and also helps when searching what's new for Python updates.

seanbudd committed:
fix up changes syntax

seanbudd committed:
fix up changes syntax

seanbudd committed:
update manifest template to drop win7/8 support

lukaszgo1 commented on sconstruct:
```diff
@@ -314,10 +314,7 @@ def NVDADistGenerator(target, source, env, for_signature):
 	buildCmd.append("-bb")
-	buildCmd.extend(("setup.py", "-vv", "build", "--build-base", buildDir.abspath,
-		"py2exe", "--dist-dir", target[0].abspath))
-	if release:
```
Shouldn't this block be restored before merging?

seanbudd replied:
No, we don't need the optimize flag. 0d6970b set the optimization level to 2 (highest) for all builds in the py2exe options

lukaszgo1 commented on user_docs/en/changes.t2t:
The fact that range of supported versions of Windows has changed should be mentioned in the change log. For 2017.4 where support for Vista/XP was dropped it was done both in the 'changes' section, and in the release blurb.

seanbudd replied:
done

seanbudd committed:
remove eoaProxy

seanbudd committed:
lower optimization level to 0

seanbudd committed:
update changes

seanbudd committed:
fix lint

lukaszgo1 commented on source/setup.py:
```diff
@@ -46,43 +42,29 @@ def determine_dll_type(self, imagename):
+		action="store_true",
+		help="enable uiAccess for the main executable"
+	)
+	partialArgs, _argslist = partialParser.parse_known_args(("--enable-uiAccess", ))
```
It looks like this is a left over from testing - I believe the intent is to parse arguments from `sys.argv` and not from the explicitly provided list.

seanbudd replied:
This is intentional - this allows us to partially parse the enable uiAccess flag, but allow py2exe to parse the rest of the parameters themselves

lukaszgo1 replied:
This has no effect that you think it does. By passing `("--enable-uiAccess", )` to the `parse_known_args` you are using the provided list of arguments rather than `sys.argv`. As a result `uiAccess` is always set to `True`, and the value provided from the CLI is ignored. Besides py2exe does not do any command line handling with the new freeze API, and looking at the arguments we're passing to the setup script most of them should be removed:
- `"-vv"` should be replaced with the `verbose"` set to 2 in the `options` dictionary provided as a parameter to `py2exe.freeze
- `"build", "--build-base", buildDir.abspath,` is a left over from when py2exe relied on setuptools, and can safely be removed
- `--dist-dir` can no longer be specified from the command line natively, so we need to either hard-code it in the call to freeze as you did in this PR, or add it as a parameter to the CLI parser ourselves, and use the CLI argument in the call to freeze.

seanbudd replied:
Ah I understand what you mean now, you were suggesting

```suggestion
	partialArgs, _argslist = partialParser.parse_known_args(sys.argv)
```

seanbudd replied:
this should be fixed now

lukaszgo1 commented on source/setup.py:
```diff
@@ -110,80 +92,60 @@ def getRecursiveDataFiles(dest,source,excludes=()):
-		},
-	],
+
+def _genManifestTemplate(shouldHaveUIAccess: bool) -> Tuple[int, int, bytes]:
```
Since we are on Python 3.11 where standard collections  can be type annotated this can be simplified to:
```suggestion
def _genManifestTemplate(shouldHaveUIAccess: bool) -> tuple[int, int, bytes]:
```

lukaszgo1 commented on source/setup.py:
```diff
@@ -110,80 +92,60 @@ def getRecursiveDataFiles(dest,source,excludes=()):
 	console=[
 		{
 			"script": os.path.join("..", "include", "nvda_dmp", "nvda_dmp.py"),
 			"uiAccess": False,
```
There is no need to specify `uiAccess` here any longer. This line can be removed.

seanbudd replied:
can you explain why? We still need to specify UI Access in the manifest for each exe

lukaszgo1 replied:
We certainly do, but we need to do this in the same way as for all other exe's i.e. by adding `_genManifestTemplate(shouldHaveUIAccess=False)` to the `other_resources` for the nvda-dmp binary.

seanbudd replied:
My mistake, this was a misunderstanding over your review comment

seanbudd commented on source/setup.py:
```diff
@@ -1,18 +1,15 @@
 import sys
-import copy
 import gettext
+from typing import Tuple
```
```suggestion
```

seanbudd committed:
fix typing

Co-authored-by: Łukasz Golonka <lukasz.golonka@mailbox.org>

seanbudd committed:
Merge remote-tracking branch 'origin/master' into py311-min

seanbudd committed:
update py2exe

seanbudd committed:
disable BRL api

seanbudd committed:
fixup uiaccess for nvda_dmp

seanbudd committed:
fix up CLI for setup py2exe

codeofdusk commented:
I've been running this for a while with almost no issues!

Sometimes though, when copying text to the clipboard in Chrome, I get:

```python traceback
Traceback (most recent call last):
  File "api.pyc", line 394, in copyToClip
  File "api.pyc", line 415, in getClipData
  File "contextlib.pyc", line 137, in __enter__
  File "winUser.pyc", line 812, in openClipboard
PermissionError: [WinError 5] Access is denied.
```

leonardder commented:
@seanbudd Would it be possible to provide a signed try build?

codeofdusk committed:
Add @codeofdusk attributions for Python 3.11 upgrade (#15609)

Fix-up of #15513 #15514 .

Description of how this pull request fixes the issue:
Add @codeofdusk attributions for Python 3.11 upgrade

seanbudd commented:
@leonardder - try build for this PR, not including brlapi support: https://ci.appveyor.com/api/buildjobs/1kkvl7p3kpu38ijm/artifacts/output%2Fnvda_snapshot_try-p311-min-29598%2Cafa0d928.exe

leonardder commented:
Thanks for the try build @seanbudd 
From my first observations, all seems to work smoothly. That said, it looks like the garbage handler kicks in more often with errors like this:

<details><summary>Log snipped</summary>
```
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Garbage collector has found one or more unreachable objects. See further warnings for specific objects.
Stack trace:
  File "nvda.pyw", line 399, in <module>
  File "core.pyc", line 813, in main
  File "wx\core.pyc", line 2262, in MainLoop
  File "wx\core.pyc", line 3427, in <lambda>
  File "core.pyc", line 746, in processRequest
  File "core.pyc", line 762, in Notify
  File "queueHandler.pyc", line 97, in pumpAll
  File "queueHandler.pyc", line 64, in flushQueue
  File "eventHandler.pyc", line 84, in _queueEventCallback
  File "eventHandler.pyc", line 322, in executeEvent
  File "eventHandler.pyc", line 355, in doPreGainFocus
  File "api.pyc", line 163, in setFocusObject
  File "treeInterceptorHandler.pyc", line 50, in update
  File "NVDAObjects\__init__.pyc", line 430, in _get_treeInterceptor
  File "treeInterceptorHandler.pyc", line 42, in getTreeInterceptor
  File "NVDAObjects\IAccessible\chromium.pyc", line 84, in __contains__
  File "comtypes\_memberspec.pyc", line 482, in __call__
  File "comtypes\client\__init__.pyc", line 47, in wrap_outparam
  File "comtypes\client\__init__.pyc", line 100, in GetBestInterface
  File "comtypes\client\dynamic.pyc", line 29, in Dispatch
  File "comtypes\automation.pyc", line 776, in GetTypeInfo
  File "monkeyPatches\comtypesMonkeyPatches.pyc", line 157, in newGetTypeInfo
  File "monkeyPatches\comtypesMonkeyPatches.pyc", line 31, in __call__
  File "monkeyPatches\comtypesMonkeyPatches.pyc", line 128, in newCpbDel
  File "garbageHandler.pyc", line 67, in notifyObjectDeletion
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x76f75b4 at 1738490>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <NVDAObjects.Dynamic_OutlineItemIa2WebIAccessibleTreeIndentNav object at 0x08C1A4F0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x76f77ac at 1738760>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x76f783c at 17381c0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x78832bc at 17385d0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x08C1A2D0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x788373c at 1738210>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x78834b4 at 1738030>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.chromium.ChromeVBufTextInfo object at 0x08C1A610>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.115) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x76f795c at 16f68f0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <NVDAObjects.Dynamic_OutlineItemIa2WebIAccessibleTreeIndentNav object at 0x08C1AF10>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x77d9c44 at 1738670>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x77d9d64 at 16f63a0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x78832bc at 16f6710>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x08C1AD10>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x788373c at 16f6da0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x78834b4 at 16f6bc0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <eventHandler._EventExecuter object at 0x08C1A870>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.chromium.ChromeVBufTextInfo object at 0x04268D30>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x77d9c8c at 5acd710>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <NVDAObjects.Dynamic_EditableTextWithAutoSelectDetectionEditorIa2WebIAccessibleEditableIndentNav object at 0x0426ED70>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x77d926c at 16f6580>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x77d953c at 16f6210>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.116) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x78832bc at 16f6530>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x0426E670>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x788373c at 5acd4e0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x78834b4 at 16f66c0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.chromium.ChromeVBufTextInfo object at 0x08C1A330>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleHypertext) ptr=0x77d95cc at 8be0490>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x76f795c at 8be0120>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <NVDAObjects.Dynamic_OutlineItemIa2WebIAccessibleTreeIndentNav object at 0x04271670>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x77d9c44 at 423ef30>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.117) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x77d9d64 at 8be0620>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.118) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessible2) ptr=0x78832bc at 16dfe40>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.118) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x04268E50>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.118) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleAction) ptr=0x788373c at 16df9e0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.118) - MainThread (17080):
Deleting unreachable object <POINTER(IAccessibleText) ptr=0x78834b4 at 16df3f0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.118) - MainThread (17080):
Deleting unreachable object <eventHandler._EventExecuter object at 0x0426ED50>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.118) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.chromium.ChromeVBufTextInfo object at 0x0426E9B0>
WARNING - garbageHandler.notifyObjectDeletion (08:34:36.118) - MainThread (17080):
Deleting unreachable object <NVDAObjects.IAccessible.chromium.ChromeVBufTextInfo object at 0x08C04D30>
ERROR - garbageHandler._collectionCallback (08:34:36.119) - MainThread (17080):
Found at least 40 unreachable objects in run
```
</details>

This was observed when arrowing through the list of chats in Teams 2.0.

I think we need to investigate what changes happend to garbage collection. It somehow looks like newer Python 3 is better in detecting abandoned objects itself.

seanbudd commented:
@leonardder @codeofdusk - these minor issues will have to be handled in a separate issue/PR, once this has been merged

michaelDCurran reviewed (approved):
In order to get maximum testing, I would recommend merging this sooner than later, acknowledging though that brlTTY support is currently not functional and a fix may still be some time away. We should *not* however branch for beta until a new brlAPI module compatible with Python 3.11 is available of course.

seanbudd committed:
Merge branch 'master' into py311-min
